src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,328 @@
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <set>
+#include <vector>
+
+#include <core/SafeConvert.hpp>
+#include <core/Algorithm.hpp>
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/system/System.hpp>
+
+#include <boost/foreach.hpp>
+#include <boost/system/error_code.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+const char * const kFileLockPrefix =
+      ".rstudio-lock-41c29";
+
+std::string pidString()
+{
+   PidType pid = system::currentProcessId();
+   return safe_convert::numberToString((long) pid);
+}
+
+std::string makeHostName()
+{
+   char buffer[256];
+   int status = ::gethostname(buffer, 255);
+   if (status)
+      LOG_ERROR(systemError(errno, ERROR_LOCATION));
+   return std::string(buffer);
+}
+
+const std::string& hostName()
+{
+   static std::string instance = makeHostName();
+   return instance;
+}
+
+std::string threadId()
+{
+   std::stringstream ss;
+   ss << boost::this_thread::get_id();
+   return ss.str();
+}
+
+std::string proxyLockFileName()
+{
+   return std::string()
+         + kFileLockPrefix
+         + "-" + hostName()
+         + "-" + pidString()
+         + "-" + threadId();
+         
+}
+
+bool isLockFileStale(const FilePath& lockFilePath)
+{
+   return LinkBasedFileLock::isLockFileStale(lockFilePath);
+}
+
+} // end anonymous namespace
+
+bool LinkBasedFileLock::isLockFileStale(const FilePath& lockFilePath)
+{
+   double seconds = s_timeoutInterval.total_seconds();
+   double diff = ::difftime(::time(NULL), lockFilePath.lastWriteTime());
+   return diff >= seconds;
+}
+
+namespace {
+
+void cleanStaleLockfiles(const FilePath& dir)
+{
+   std::vector<FilePath> children;
+   Error error = dir.children(&children);
+   if (error)
+      LOG_ERROR(error);
+   
+   BOOST_FOREACH(const FilePath& filePath, children)
+   {
+      if (boost::algorithm::starts_with(filePath.filename(), kFileLockPrefix) &&
+          isLockFileStale(filePath))
+      {
+         Error error = filePath.removeIfExists();
+         if (error)
+            LOG_ERROR(error);
+      }
+   }
+}
+
+class LockRegistration : boost::noncopyable
+{
+public:
+   
+   void registerLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.insert(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void deregisterLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.erase(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void refreshLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.setLastWriteTime();
```

Comment:
One thing @jmcphers pointed out is that these kind of file metadata changes can potentially be cached on NFS systems (for up to a minute?). I wonder if we can / need to do anything to guarantee a more immediate update of the write time?